### ISSUE TITLE
False positive ``invalid-repr-returned`` with ``@overload``

### ISSUE THREAD

**Issue by MarcoGorelli** [PyCQA/pylint#7624]
### Bug description
```python
from __future__ import annotations

from typing import overload

class Foo:
    def __repr__(self) -> str:
        return self.foo()

    @overload
    def foo(self) -> str:
        ...

    def foo(self) -> str:
        return 'foo'
```

### Command used
I just put the above in a file `f.py` and ran
```shell
pylint --disable all --enable invalid-repr-returned f.py
```

### Pylint output

```shell
************* Module f
f.py:6:4: E0306: __repr__ does not return str (invalid-repr-returned)

-------------------------------------------------------------------
Your code has been rated at 4.44/10 (previous run: 10.00/10, -5.56)
```

### Expected behavior

I don't think this should raise

### Pylint version

```shell
pylint 2.15.4
astroid 2.12.11
Python 3.8.13 | packaged by conda-forge | (default, Mar 25 2022, 06:04:10) 
[GCC 10.3.0]
```

### OS / Environment

Ubuntu (via WSL2)

**Comment by jacobtylerwalls** [comment]
Thanks for the report.

@Pierre-Sassoulas I suggest closing as a likely duplicate of PyCQA/astroid#1015 as long as we note this test case for pylint.
